scripts/generate_matrix.py

```diff
@@ -0,0 +1,273 @@
+
+
+def get_dependency_minimum_version(package: str, dependency: str) -> Version:
+    """
+    Get the minimum version of a dependency required by the specified package.
+
+    Args:
+        package (str): The name of the package.
+        dependency (str): The name of the dependency.
+
+    Returns:
+        Version: The minimum required version of the dependency.
+
+    Raises:
+        ValueError: If the package does not list any requirements or the dependency is not found.
+
+    """
+    requirements = importlib.metadata.requires(package)
+    if not requirements:
+        raise ValueError(f"No requirements for {package}")
+
+    try:
+        # Find the first requirement that matches the dependency and is not an extra
+        dependency_req = next(req for req in requirements if req.startswith(dependency) and " extra " not in req)
+    except StopIteration as e:
+        raise ValueError(f"Dependency {dependency} not found in requirements for {package}") from e
+
+        # Extract the version specifier (e.g., ">=1.21.0")
+    version_specifier = next(
+        (ver for ver in dependency_req.split(",") if ">" in ver),
+        None,
+    )
+    if version_specifier is None:
+        raise ValueError(f"No version specifier found for dependency {dependency} in {package}")
+
+    # Remove dependency name and comparison operator to get the version
+    version_str = version_specifier.replace(dependency, "").replace(">=", "").replace(">", "")
+    return parse(version_str)
+
+
+def get_available_python_versions(
+    min_version: Version | None = None,
+    max_version: Version | None = None,
+    pre_releases: bool = False,
+) -> list[Version]:
+    """
+    Get a list of available Python versions from GitHub Actions' Python Versions Manifest.
+
+    Args:
+        min_version (Version | None): The minimum Python version to include in the list.
+        max_version (Version | None): The maximum Python version to include in the list.
+        pre_releases (bool): Whether to include pre-release versions.
+
+    Returns:
+        list[Version]: A list of available Python versions satisfying the specified criteria.
+
+    Raises:
+        urllib.error.URLError: If fetching data fails.
+
+    """
+    data: list[Release] = typing.cast(
+        "list[Release]",
+        fetch_json("https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json"),
+    )
+
+    versions: dict[tuple[int, int], Version] = {}
+
+    for release in data[::-1]:
+        version = parse(release["version"])
+
+        if version.is_prerelease and not pre_releases:
+            continue
+
+        if min_version and version < min_version:
+            continue
+        if max_version and version > max_version:
+            continue
+
+        versions[version.major, version.minor] = version
+
+    return sorted(versions.values())
+
+
+@lru_cache(maxsize=CACHE_SIZE)  # type: ignore[no-any-expr]
+def fetch_json(url: str) -> object:
```

Comment:
returning `Any` avoids having to `typing.cast` in `get_available_package_versions` (`Any` is still evil, though)

Reply:
It seems to make mypy unhappy though... so it doesn't help :(

Reply:
yea, `based{mypy,pyright}` declared a holy war against `Any` (and therefore gradual typing as a whole, for which they recommend their "baseline" feature as an alternative), so you'd have to ignore some errors in that case, so perhaps a `cast` is indeed cleaner